hydra-main/src/main/java/com/addthis/hydra/job/spawn/balancer/HostCandidateIterator.java

```diff
@@ -0,0 +1,129 @@
+        if(replicas <= 0) {
+            return new ArrayList<>();
+        }
+
+        Collection<HostLocation> locations = new ArrayList<>();
+        // get current host locations used by live/replicas
+        for (String replicaHostId : task.getAllTaskHosts()) {
+            if ((replicaHostId != null) && !replicaHostId.equals(excludeHostUuid)) {
+                locations.add(hostManager.getHostState(replicaHostId).getHostLocation());
+            }
+        }
+
+        List<String> chosenHostIds = new ArrayList<>(replicas);
+        // create each replica
+        for (int i = 0; i < replicas; i++) {
+            double minScoreSoFar = Double.MAX_VALUE;
+            WithScore<HostState> bestHost = null;
+            for (WithScore<HostState> hostAndScore : sortedHosts) {
+                HostState host = hostAndScore.element;
+                boolean excluded = host.getHostUuid().equals(excludeHostUuid);
+                if (excluded || !host.canMirrorTasks() || (isReplica && !balancer.okToPutReplicaOnHost(host, task))) {
+                    continue;
+                }
+                HostLocation hostLocation = host.getHostLocation();
+                int score = 0;
+                // HostLocation further away from task and replica locations gets lower score
+                for (HostLocation location : locations) {
+                    score += hostLocation.assignScoreByHostLocation(location);
+                }
+                double avgScore = locations.isEmpty()? 0 : (double) score / (double) locations.size();
+                if (avgScore == (double) this.minScore) {
+                    bestHost = hostAndScore;
+                    break;
+                }
+                if (avgScore < minScoreSoFar) {
+                    // should cache this host as it's further away than the previous best host we've seen
+                    bestHost = hostAndScore;
+                    minScoreSoFar = avgScore;
+                }
+            }
+            if (bestHost != null) {
+                this.sortedHosts.remove(bestHost);
+                this.sortedHosts.add(new WithScore<>(bestHost.element, bestHost.score + (double) this.taskScoreIncrement));
+                locations.add(bestHost.element.getHostLocation());
+                chosenHostIds.add(bestHost.element.getHostUuid());
+            }
+        }
+        return chosenHostIds;
+    }
+
+    /**
+     * Get new replica host
+     */
+    public List<String> getNewReplicaHosts(int replicas, JobTask task) {
+        return this.getNewReplicaHosts(replicas, task, null, true);
+    }
+
+    /**
+     * Get new replica host for a rebalancing task
+     * note: this is not used at the moment, but saved for a later date when we rewrite spawnbalancer
+     */
+    public List<String> getNewReplicaHosts(JobTask task, @Nullable String excludeHost) {
+        return this.getNewReplicaHosts(1, task, excludeHost, true);
+    }
+
+}
```

Comment:
it looks like it can return an empty list, while a better approach would be it can always give a list of size `replicas`, so that it will not block things

Reply:
An empty list is returned only when `replicas <= 0`; a list of size `replicas` is returned otherwise - am I missing some line of code where we return an empty list?